### DISCUSSION TITLE
Red Topics not working

### DISCUSSION THREAD

**Discussion by shu-bamma** [larics/icuas22_competition#40] [Q&A]
Hello,
The topics titled red/mavros like velocity setpoints and setpoint raw do not work on publishing values to them although they come up in rostopic list
What seems to be the issue here??

**Comment by fpetric** [comment]
Can you confirm that `/red/pid_cascade_node` is running? Also, please check if ardupilot gazebo plugin has been loaded successfully

**Comment by lmark1** [comment]
The only setpoint topic that mavros accepts is the ```/red/mavros/setpoint_raw/attitude```. You are not meant to design a controller, but rather navigate the UAV around the environment using the topic described [here](https://github.com/larics/icuas22_competition#controlling-the-uav).

**Reply by fpetric** [reply]
@shu-bamma as @lmark1 said, it wasn't the aim of the competition to design "low-level" controllers, but if you really want to, see [this discussion](https://github.com/larics/icuas22_competition/discussions/34#discussioncomment-2360114).

**Reply by shu-bamma** [reply]
@fpetric  @lmark1  What if We want to publish velocity commands in the drones local frame? which topic can we use so that it is consistent with the hardware challenge?

**Reply by fpetric** [reply]
@lmark1 may correct me, but you can feed forward the velocity commands (and acceleration too) through the trajectory reference, you just need to set `ff_gain` for the signals you want to control to 1.0 in the [position control config file](https://github.com/larics/icuas22_competition/blob/86799e1b00bd91ff02ac0f50b603586c28bfc222/startup/challenge/custom_config/position_control_custom.yaml). I believe you need to transform the commands from the drone local frame to world frame before sending, but that should not be too much of a problem (@lmark1 please confirm). 

As far as I am aware, that is the only way to control velocities and accelerations that will be available on the platform for the finals.

**Reply by IllierZer** [reply]
> @lmark1 may correct me, but you can feed forward the velocity commands (and acceleration too) through the trajectory reference, you just need to set `ff_gain` for the signals you want to control to 1.0 in the [position control config file](https://github.com/larics/icuas22_competition/blob/86799e1b00bd91ff02ac0f50b603586c28bfc222/startup/challenge/custom_config/position_control_custom.yaml). I believe you need to transform the commands from the drone local frame to world frame before sending, but that should not be too much of a problem (@lmark1 please confirm).
> 
> As far as I am aware, that is the only way to control velocities and accelerations that will be available on the platform for the finals.

@lmark1 @fpetric setting the ff_gain in the given .yaml file doesnt seem to help, we particularly want to control only the velocity and not the position, basically we are looking for the functionality provided by the topic /mavros/setpoint_velocity/cmd_vel where we can publish velocity commands to the drone in the drone's local frame without being concerned about the position whatsoever.

**Reply by lmark1** [reply]
Hello @IllierZer ,

The current controller implementation does not allow pure velocity control.
Setting the ff_gain only feedwords the velocities to the Cascade PID controller, but at the end of the day it's still a position controller.

You can however "trick" the controller by publishing position controls relative to ```/red/carrot/pose``` in the direction you so desire.
For example, if you wish to move the drone forward in its body axis simply take a Vector(1,0,0), transform it to the UAV body frame, add it to ```/red/carrot/pose``` and publish it either to ```/red/position_hold/trajectory``` or ```/red/tracker/input_pose```. 

That way you can essentially control the drones relative position, without worrying about absolute measurements.

Best regards,
Lovro
